webrtc/browser-to-browser.md

```diff
@@ -0,0 +1,103 @@
+
+## Motivation
+
+1. **Hole punching in the browser**: Enable two browsers or a browser and a server node to connect even though one or both are behind a NAT / firewall.
+
+   TODO: Doucment use-case where A is a browser and B is a non-browser but behind firewall and/or NAT.
+
+## Connection Establishment
+
+### Browser to Browser
+
+Scenario: Browser _A_ wants to connect to Browser node _B_ with the help of server node _R_.
+Both _A_ and _B_ can not listen for incoming connections due to the restriction of the browser platform and being behind a NAT and/or firewall.
+
+TODO: Define which node on a relayed connection is _A_ and which one is _B_, i.e. which one initiates (offer) and which one responds (answer).
+
+1. _A_ and _B_ establish a relayed connection through some protocol, e.g. the Circuit Relay v2 protocol.
+   Note that further steps depend on the relayed connection to be authenticated, i.e. that data send on the relayed connection can be trusted.
+
+2. _A_ creates an `RTCPeerConnection`.
+   See [#STUN] on what STUN servers to configure at creation time.
+   _A_ creates an SDP offer via `RTCPeerConnection.createOffer()`.
+   _A_ initiates the signaling protocol to _B_ via the relayed connection from (1), see [#Signaling protocol] and sends the offer to _B_.
+
+3. _B_ creates an `RTCPeerConnection`.
+   Again see [#STUN] on what STUN servers to configure at creation time.
+   _B_ receives _A_'s offer send in (2) via the signaling protocol stream and provides the offer to its `RTCPeerConnection` via `RTCPeerConnection.setRemoteDescription`.
+   _B_ then creates an answer via `RTCPeerConnection.createAnswer` and sends it to _A_ via the existing signaling protocol stream (see [#signaling protocol]).
+
+4. _A_ receives _B_'s answer via the signaling protocol stream and sets it locally via `RTCPeerConnection.setRemoteDescription`.
+
+5. _A_ and _B_ send their local ICE candidates via the existing signaling protocol stream.
+   Both nodes continuously read from the stream, adding incoming remote candidates via `RTCPeerConnection.addIceCandidate()`.
+
+6. On successful establishment or failure of the direct connection, _A_ and _B_ close the signaling protocol stream.
+   TODO: Is there value in retrying on failure?
+
+7. Messages on `RTCDataChannel`s on the established `RTCPeerConnection` are framed using the message framing mechanism described in [Multiplexing](#multiplexing).
+
+The above browser-to-browser WebRTC connection establishment replaces the existing [libp2p WebRTC star](https://github.com/libp2p/js-libp2p-webrtc-star) and [libp2p WebRTC direct](https://github.com/libp2p/js-libp2p-webrtc-direct) protocols.
+
+## STUN
+
+TODO: Specify
+
+## Signaling protocol
+
+The protocol id is `/webrtc-direct`.
+Messages are sent prefixed with the message length in bytes, encoded as an unsigned variable length integer as defined by the [multiformats unsigned-varint spec][uvarint-spec].
+
+``` protobuf
+syntax = "proto2";
```

Comment:
Consider proto3: https://github.com/libp2p/specs/issues/465

Reply:
I don't think that decision should happen for each specification individually. Until https://github.com/libp2p/specs/issues/465 is resolved, I think all specifications should follow the existing consensus, namely to use proto2. See https://github.com/libp2p/specs/#protocols.

I am fine with libp2p moving to proto3 (see https://github.com/libp2p/specs/issues/465#issuecomment-1272291524). I myself have not pushed https://github.com/libp2p/specs/issues/465 forward, as I don't think the benefits of moving to proto3 are worth prioritizing the move over other work (e.g. this specification).

Reply:
https://github.com/libp2p/specs/pull/506 there you go :)

I think this message is compatible with proto3, so it would be a one line change.

Reply:
Moved to proto3. See https://github.com/libp2p/specs/pull/497/commits/2408db97f0d8883916486d08ae2720c286fa946b.